src/grpc/packet_router/hpr_packet_router_service.erl

```diff
@@ -114,6 +113,93 @@ register(PubKeyBin) ->
+    {EnvDown, StreamState1}.
+
+-spec handle_packet(PacketUp :: hpr_packet_up:packet(), StreamState0 :: grpcbox_stream:t()) ->
+    {ok, grpcbox_stream:t()}.
+handle_packet(PacketUp, StreamState) ->
+    HandlerState = grpcbox_stream:stream_handler_state(StreamState),
+    SessionKey = HandlerState#handler_state.session_key,
+    _ = erlang:spawn_opt(hpr_routing, handle_packet, [PacketUp, SessionKey], [{fullsweep_after, 0}]),
+    {ok, StreamState}.
+
+-spec handle_register(Reg :: hpr_register:register(), StreamState0 :: grpcbox_stream:t()) ->
+    {ok, grpcbox_stream:t()}
+    | {ok, hpr_envelope_down:envelope(), grpcbox_stream:t()}
+    | {stop, grpcbox_stream:t()}.
+handle_register(Reg, StreamState0) ->
+    PubKeyBin = hpr_register:gateway(Reg),
+    lager:md([{gateway, hpr_utils:gateway_name(PubKeyBin)}]),
+    case hpr_register:verify(Reg) of
+        false ->
+            lager:warning("failed to verify register"),
+            #handler_state{started = Started} = grpcbox_stream:stream_handler_state(
+                StreamState0
+            ),
+            _ = hpr_metrics:observe_grpc_connection(?MODULE, Started),
+            {stop, StreamState0};
+        true ->
+            ok = ?MODULE:register(PubKeyBin),
+            case hpr_register:session_capable(Reg) of
+                true ->
+                    {EnvDown, StreamState1} = create_session_offer(StreamState0),
+                    {ok, EnvDown, StreamState1};
+                false ->
+                    {ok, StreamState0}
+            end
+    end.
+
+-spec handle_session_init(Reg :: hpr_session_init:init(), StreamState0 :: grpcbox_stream:t()) ->
+    {ok, grpcbox_stream:t()} | {stop, grpcbox_stream:t()}.
+handle_session_init(SessionInit, StreamState0) ->
+    case hpr_session_init:verify(SessionInit) of
+        false ->
+            lager:warning("failed to verify session init"),
+            {stop, StreamState0};
+        true ->
+            HandlerState0 = grpcbox_stream:stream_handler_state(StreamState0),
+            case hpr_session_init:nonce(SessionInit) == HandlerState0#handler_state.nonce of
+                false ->
+                    lager:warning("nonce did not match ~s ~s", [
+                        hpr_utils:bin_to_hex_string(hpr_session_init:nonce(SessionInit)),
+                        hpr_utils:bin_to_hex_string(HandlerState0#handler_state.nonce)
+                    ]),
+                    {ok, StreamState0};
+                true ->
+                    lager:debug("session init ~s", [
+                        hpr_utils:bin_to_hex_string(hpr_session_init:nonce(SessionInit))
+                    ]),
+                    StreamState1 = grpcbox_stream:stream_handler_state(
+                        StreamState0, HandlerState0#handler_state{
+                            session_key = hpr_session_init:session_key(SessionInit)
+                        }
+                    ),
+                    ok = schedule_session_reset(),
```

Comment:
If a gw sends a bunch of session_inits, it can start indefinitely many timers.

Reply:
will fix